R/module_bookmark_manager.R

```diff
@@ -0,0 +1,222 @@
+    app_session$onRestored(function(state) {
+      # Restore filter state.
+      logger::log_trace("bookmark_manager_srv@onRestored: restoring filter state")
+      snapshot <- state$values$filter_state_on_bookmark
+      snapshot_state <- as.teal_slices(snapshot)
+      mapping_unfolded <- unfold_mapping(attr(snapshot_state, "mapping"), names(datasets))
+      mapply(
+        function(filtered_data, filter_ids) {
+          filtered_data$clear_filter_states(force = TRUE)
+          slices <- Filter(function(x) x$id %in% filter_ids, snapshot_state)
+          filtered_data$set_filter_state(slices)
+        },
+        filtered_data = datasets,
+        filter_ids = mapping_unfolded
+      )
+      slices_global(snapshot_state)
+      # Restore snapshot history and bookmark history.
+      logger::log_trace("bookmark_manager_srv@onRestored: restoring snapshot and bookmark history")
+      snapshot_history(state$values$snapshot_history)
+      bookmark_history(state$values$bookmark_history)
+    })
+
+    ns <- session$ns
+
+    # Store input states.
+    bookmark_history <- reactiveVal({
+      list()
+    })
+
+    # Bookmark current input state - name bookmark.
+    observeEvent(input$bookmark_add, {
+      logger::log_trace("bookmark_manager_srv: bookmark_add button clicked")
+      showModal(
+        modalDialog(
+          textInput(ns("bookmark_name"), "Name the bookmark", width = "100%", placeholder = "Meaningful, unique name"),
+          footer = tagList(
+            actionButton(ns("bookmark_accept"), label = "Accept", icon = icon("thumbs-up")),
+            modalButton(label = "Cancel", icon = icon("thumbs-down"))
+          ),
+          size = "s"
+        )
+      )
+    })
+
+    # Initiate bookmarking with normal action button b/c `bookmarkButton` may not work on Windows.
+    observeEvent(input$bookmark_accept, {
+      app_session$doBookmark()
+    })
+
+    # Create UI elements and server logic for the bookmark table.
+    # Divs are tracked for a slight speed margin.
+    divs <- reactiveValues()
+
+    observeEvent(bookmark_history(), {
+      logger::log_trace("bookmark_manager_srv: bookmark history changed, updating bookmark list")
+      lapply(names(bookmark_history()), function(s) {
+        id_rowme <- sprintf("rowme_%s", make.names(s))
+
+        # Create a row for the bookmark table.
+        if (!is.element(id_rowme, names(divs))) {
+          divs[[id_rowme]] <- div(
+            class = "manager_table_row",
+            a(h5(s), title = "go to bookmark", href = bookmark_history()[[s]], target = "blank")
+          )
+        }
+      })
+    })
+
+    # Create table to display list of bookmarks and their actions.
+    output$bookmark_list <- renderUI({
+      rows <- lapply(rev(reactiveValuesToList(divs)), function(d) d)
+      if (length(rows) == 0L) {
+        div(
+          class = "manager_placeholder",
+          "Bookmarks will appear here."
+        )
+      } else {
+        rows
+      }
+    })
```

Comment:
I think we might be in need for more metadata for individual bookmark. What you did is an open text `bookmark_name` which is a good start already. 
If this functionality is meant to be shared across the users it would be good to preserve author somehow.
Also, when dealing with _large_ number of bookmarks it would be great to have a timestamp (and allow to sort this table to have the newest at the top).
Maybe also an open text _description_ to use the existing _name_ properly and throw everything else in there.

We can probably go with more. I am constantly thinking about _tags_ but this would be fine to add in time. I feel that the ones I described above are important for a start.